control/metrics.go

```diff
@@ -706,3 +546,126 @@ func addStandardAndWorkflowTags(m core.Metric, allTags map[string]map[string]str
+	// 	"intel", "mock", "(host0;host1)", "(baz;bar)"
+	//
+	// the following 4 namespaces will be returned:
+	// 	"intel", "mock", "host0", "baz"
+	// 	"intel", "mock", "host1", "baz"
+	// 	"intel", "mock", "host0", "bar"
+	// 	"intel", "mock", "host1", "bar"
+
+	matchedItems := make(map[int][]string)
+	numOfPossibleCombinations := 1
+
+	for index, element := range incomingNs.Strings() {
+		match := []string{}
+		if ok, tupleItems := containsTuple(element); ok {
+			match = tupleItems
+		} else {
+			match = []string{element}
+		}
+		// store matched items under current index of incoming namespace element
+		matchedItems[index] = append(matchedItems[index], match...)
+
+		// number of possible combinations increases N=len(match) times
+		numOfPossibleCombinations = numOfPossibleCombinations * len(match)
+	}
+
+	//prepare slice for returned namespaces (results of tuple find)
+	returnedNss := make([]core.Namespace, numOfPossibleCombinations)
+
+	// initialize each of returned namespaces as a copy of incoming namespace
+	// (copied original value, name and description of their elements)
+	for i := 0; i < numOfPossibleCombinations; i++ {
+		returnedNs := make([]core.NamespaceElement, len(incomingNs.Strings()))
+		copy(returnedNs, incomingNs)
+		returnedNss[i] = returnedNs
+	}
+	// set appropriate value to namespace's elements
+	for index, items := range matchedItems {
+		for i := range returnedNss {
+			// retrieve the matched item (when 'i' exceeds the number of matched items, start from beginning)
+			item := items[i%len(items)]
+			returnedNss[i][index].Value = item
+		}
+	}
+	return returnedNss
+}
+
+// specifyInstanceOfDynamicMetric returns specified namespace of incoming cataloged metric's namespace based on requested metric namespace
+func specifyInstanceOfDynamicMetric(catalogedNamespace core.Namespace, requestedNamespace core.Namespace) core.Namespace {
+	specifiedNamespace := make(core.Namespace, len(catalogedNamespace))
+	copy(specifiedNamespace, catalogedNamespace)
+
+	_, indexes := catalogedNamespace.IsDynamic()
+
+	for _, index := range indexes {
+		if len(requestedNamespace) > index {
+			// use namespace's element of requested metric declared in task manifest
+			// to specify a dynamic instance of the cataloged metric
+			specifiedNamespace[index].Value = requestedNamespace[index].Value
+		}
+	}
+	return specifiedNamespace
+}
+
+// validateMetricNamespace validates metric namespace in terms of containing properly defined dynamic elements, not ending with an asterisk
+// and not contain elements which might be erroneously recognized as a tuple
+func validateMetricNamespace(ns core.Namespace) error {
+	value := ""
+	for _, i := range ns {
+		// A dynamic element requires the name while a static element does not.
+		if i.Name != "" && i.Value != "*" {
+			return errorMetricStaticElementHasName(i.Value, i.Name, ns.String())
+		}
+		if i.Name == "" && i.Value == "*" {
+			return errorMetricDynamicElementHasNoName(i.Value, ns.String())
```

Comment:
Yes, definitely - I've just removed this `validateMetricNamespace` to the end of file, so that is not my changes (it's existing code, see [snap/blob/master/control/metrics.go#L376](https://github.com/intelsdi-x/snap/blob/master/control/metrics.go#L376)), but it should be changed to `isDynamic()`. Thanks for notice that!